iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -49,458 +50,528 @@ namespace mlir {
+// typically meant to be to target scratchspace memory.
+//
+// The algorithm has two phases
+// and analysis phase and a tranformation phase.
+//
+// - The analysis phase walks the function and organizes relevant tensors
+//   (tensors that need to be converted to memrefs) into equivalence clases. Two
+//   tensors are part of the same equivalence class if they can eventually be
+//   mapped to the same memref. This allows determining which operations can use
+//   the buffer provided for the outputs to compute the results in place.
+// - The transformation phase walks the function again and inserts corresponding
+//   memref operations. The tensor operations are still kept around since the
+//   analysis driving the transformation is based on the tensor values.
+//   - Converting tensor operations to memref operations when all operands use
+//     either buffers that are inputs to the dispatch or are allocated
+//     temporarily within the dispatch region can be achieved by a a
+//     straight-forward walk.
+//   - Reusing memref for the result of the dispatch for operations is more
+//     involved and explained below.
+//
 //===----------------------------------------------------------------------===//
 
-static MemRefType getMemrefTypeForTensor(RankedTensorType tensorType,
-                                         ArrayRef<AffineMap> layout = {},
-                                         unsigned memorySpace = 0) {
-  return MemRefType::get(tensorType.getShape(), tensorType.getElementType(),
-                         layout, memorySpace);
+//===----------------------------------------------------------------------===//
+// ANALYSIS TO COMPUTE EQUIVALENCE SETS
+//
+// These functions compute the equivalence relationships between all tensors in
+// the program. Two tensors are equivalent if they are to be mapped to the same
+// buffer. For every operation, based on the operation semantics the result of
+// the operation can reuse the buffer for an operand of the operation. This
+// information is captured by adding these two tensors to the same equivalence
+// class. Eventually the result of the dispatch tensor is added to some
+// equivalence set. All tensors in that equivalence set can reuse the result
+// buffer and compute the values in place. You can add tensors to equivalence
+// setonly if
+// - They have a single use
+// - They are derived from a read-only buffer.
+//
+//===----------------------------------------------------------------------===//
+
+using UnionFindSet = llvm::EquivalenceClasses<void *>;
+
+/// Walks the use-def chain and see if this value comes from a read-only tensor.
+static bool isFromReadOnlyTensor(Value v) {
+  auto definingOp = v.getDefiningOp();
+  if (!definingOp) return false;
+  return TypeSwitch<Operation *, bool>(definingOp)
+      .Case<ConstantOp>([&](ConstantOp constantOp) { return true; })
+      .Case<linalg::TensorReshapeOp>([&](linalg::TensorReshapeOp reshapeOp) {
```

Comment:
add a case to walk `TensorCastOp` as well ?

Reply:
Maybe. Right now there is no need for that. We can add it later on.

Reply:
I could be wrong bug my understanding If this is similar to the forward walk in the old code, that might cause an `alloca` + `copy` for cast